src/lib/transition_handler/catchup_scheduler.ml

```diff
@@ -84,37 +97,51 @@ module Make (Inputs : Inputs.S) = struct
     in
     let make_timeout () =
       Time.Timeout.create t.time_controller timeout_duration ~f:(fun _ ->
-          (* it's ok to create a new thread here because the thread essentially does no work *)
           don't_wait_for (Writer.write t.catchup_job_writer transition) )
     in
-    Hashtbl.update t.timeouts parent_hash ~f:(function
-      | None -> [(transition, make_timeout ())]
-      | Some entries ->
+    Hashtbl.update t.collected_transitions parent_hash ~f:(function
+      | None ->
+          cancel_child_timeout t hash ;
+          Hashtbl.add_exn t.collected_transitions ~key:hash ~data:[] ;
+          Hashtbl.add_exn t.timeouts ~key:parent_hash ~data:(make_timeout ()) ;
+          [transition]
+      | Some collected_transitions ->
           if
-            List.exists entries ~f:(fun (trans, _) ->
-                State_hash.equal hash (With_hash.hash trans) )
+            List.exists collected_transitions ~f:(fun collected_transition ->
+                State_hash.equal hash @@ With_hash.hash collected_transition )
           then (
             Logger.info t.logger
               !"Received request to watch transition for catchup that already \
                 was being watched: %{sexp: State_hash.t}"
               hash ;
-            entries )
-          else (transition, make_timeout ()) :: entries )
+            collected_transitions )
+          else (
+            cancel_child_timeout t hash ;
+            Hashtbl.add_exn t.collected_transitions ~key:hash ~data:[] ;
+            transition :: collected_transitions ) )
 
-  let rec extract t (transition, timeout) =
-    Time.Timeout.cancel t.time_controller timeout () ;
+  let rec extract t transition =
     let successors =
       Option.value ~default:[]
-        (Hashtbl.find t.timeouts (With_hash.hash transition))
+        (Hashtbl.find t.collected_transitions (With_hash.hash transition))
     in
     Rose_tree.T (transition, List.map successors ~f:(extract t))
 
   let notify t ~transition =
-    Option.iter
-      (Hashtbl.find t.timeouts (With_hash.hash transition))
-      ~f:(fun entries ->
-        Hashtbl.remove t.timeouts (With_hash.hash transition) ;
-        let transition_branches = List.map entries ~f:(extract t) in
+    let hash = With_hash.hash transition in
+    Hashtbl.find_and_call t.timeouts hash
+      ~if_found:(fun timeout ->
+        Time.Timeout.cancel t.time_controller timeout () )
+      ~if_not_found:(fun _ ->
+        Logger.info t.logger
+          !"Failed to find the timeout for %{sexp: State_hash.t}"
+          hash ) ;
+    Option.iter (Hashtbl.find t.collected_transitions hash)
```

Comment:
I do not think you are not deleting any transitions in collected_transitions after turning them into breadcrumbs. Can you do that?